### PR TITLE
Long lived rule cache and rule validation hooks

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?
Avoid re-loading rules that are in the rule cache by introducing long-lived `LogicCache`. The `LogicCache` contains the prior`TypeHinterCache` and the new `RuleCache`. We also introduce a scan for retrieving all rule structures that are stored in the schema graph, and leave hooks for rule validation points. Rule validation now occurs on both rule `put` and on schema transaction `commit`

## What are the changes implemented in this PR?
* Introduce `LogicCache`, containing `RuleCache` and `TypeHinterCache`
* Introduce a scan for all rules in `SchemaGraph`. We will improve this by indexing rules by types they mention in the future, so we can look up rules that are relevant to a starting rule and the types it can infer
* Introduce points for rule validation: at `put` time, and re-validation at schema transaction commit time (in case the schema has chanced since writing the rule the first time)
* rename `Grakn.Transaction.logics()` to `Grakn.Transaction.logic()`
* Simplify `Rule` by removing `ConceptManager`

### CODE REVIEW

File: graph/SchemaGraph.java [hunk 1]
```diff
@@ -166,6 +164,13 @@ public TypeVertex rootRoleType() {
         return getType(ROLE.label(), ROLE.scope());
     }
 
+    public ResourceIterator<RuleStructure> rules() {
```
Comment (flyingsilverfin) [inline graph/SchemaGraph.java:167]: scan index for all rules, then create `RuleStructure` objects from them

File: logic/Rule.java [hunk 2]
```diff
@@ -165,57 +146,50 @@ public final int hashCode() {
...
-    private void validateLabelsExist() {
-        Stream<Label> whenPositiveLabels = getTypeLabels(when.variables().stream());
-        Stream<Label> whenNegativeLabels = getTypeLabels(when.negations().stream().flatMap(this::negationVariables));
-        Stream<Label> thenLabels = getTypeLabels(then.stream().flatMap(constraint -> constraint.variables().stream()));
-        Set<String> invalidLabels = invalidLabels(Stream.of(whenPositiveLabels, whenNegativeLabels, thenLabels).flatMap(Function.identity()));
-        if (!invalidLabels.isEmpty()) {
-            throw GraknException.of(TYPES_NOT_FOUND, getLabel(), String.join(", ", invalidLabels));
-        }
+    void validateCycles() {
+        // TODO detect negated cycles in the rule graph
```
Comment (flyingsilverfin) [inline logic/Rule.java:155]: kasper's slot in

File: logic/LogicManager.java [hunk 3]
```diff
@@ -18,43 +18,133 @@
...
-        // TODO detect negated cycles in the rule graph after inserting this rule and updating a rule graph
-
-        return rule;
+        RuleStructure structure = graphMgr.schema().getRule(label);
+        if (structure != null) {
+            // overwriting a rule means we purge it and re-create the rule
+            structure.delete();
+            logicCache.rule().invalidate(label);
+        }
+        return logicCache.rule().get(label, l -> Rule.of(graphMgr, conceptMgr, this, label, when, then));
```
Comment (jmsfltchr) [inline logic/LogicManager.java:68]: We have ensured the rule has been deleted if it exists; So this last line is creating a rule, putting it in the cache, and storing it to the graph, and then returning it from the cache?

File: logic/LogicManager.java [hunk 4]
```diff
@@ -18,43 +18,133 @@
...
+            return logicCache.rule().get(structure.label(), l -> Rule.of(this, structure));
+        }
         return null;
     }
 
-    public TypeHinter typeHinter() {
+    public ResourceIterator<Rule> rules() {
+        return graphMgr.schema().rules().map(struct -> {
+            Rule rule = logicCache.rule().getIfPresent(struct.label());
+            if (rule == null) rule = logicCache.rule().get(struct.label(), l -> Rule.of(this, struct));
```
Comment (jmsfltchr) [inline logic/LogicManager.java:84]: We check nulls everywhere in the storage layer rather than use `Optional<>`s?

**Review by jmsfltchr — APPROVED** [review]

File: logic/LogicManager.java (hunk 3 above)
Comment (flyingsilverfin) [reply logic/LogicManager.java:68]: yup - and atomically!

File: logic/LogicManager.java (hunk 4 above)
Comment (flyingsilverfin) [reply logic/LogicManager.java:84]: yep

Comment (flyingsilverfin) [reply logic/LogicManager.java:84]: because rocks can return null values for a key

Comment (haikalpribadi) [reply logic/LogicManager.java:84]: how can RocksDB return null values for a key, actually, @flyingsilverfin ?

**Review by haikalpribadi — APPROVED** [review]

**Pull request merged**